ipa-core/src/report.rs

```diff
@@ -468,73 +671,205 @@ where
-            &buf[sizeof_matchkey + ts_sz..sizeof_matchkey + ts_sz + bk_sz],
+            &buf[mk_sz + ts_sz..mk_sz + ts_sz + bk_sz],
         ));
         let trigger_value = Replicated::<TV>::deserialize(GenericArray::from_slice(
-            &buf[sizeof_matchkey + ts_sz + bk_sz..sizeof_matchkey + ts_sz + bk_sz + tv_sz],
+            &buf[mk_sz + ts_sz + bk_sz..mk_sz + ts_sz + bk_sz + tv_sz],
         ));
         let is_trigger = Replicated::<Boolean>::deserialize(GenericArray::from_slice(
-            &buf[sizeof_matchkey + ts_sz + bk_sz + tv_sz
-                ..sizeof_matchkey + ts_sz + bk_sz + tv_sz + sizeof_eventtype],
+            &buf[mk_sz + ts_sz + bk_sz + tv_sz..mk_sz + ts_sz + bk_sz + tv_sz + it_sz],
         ));
+        let epoch = u16::from_le_bytes([
+            buf[mk_sz + ts_sz + bk_sz + tv_sz + it_sz],
+            buf[mk_sz + ts_sz + bk_sz + tv_sz + it_sz + 1usize],
+        ]);
+        // unwrap is safe since it deserializes serialized valid strings
+        let site_domain = String::from_utf8(
+            buf[mk_sz + ts_sz + bk_sz + tv_sz + it_sz + 2usize..Self::Size::USIZE].to_vec(),
+        )
+        .unwrap();
         Self {
             match_key,
             is_trigger,
             breakdown_key,
             trigger_value,
             timestamp,
+            epoch,
+            site_domain,
         }
     }
 }
 
+impl<BK, TV, TS> OprfReport<BK, TV, TS>
+where
+    BK: SharedValue,
+    TV: SharedValue,
+    TS: SharedValue,
+    Replicated<BK>: Serializable,
+    Replicated<TV>: Serializable,
+    Replicated<TS>: Serializable,
+{
+    const TS_OFFSET: usize = 0;
+    const BK_OFFSET: usize = Self::TS_OFFSET + <Replicated<TS> as Serializable>::Size::USIZE;
+    const TV_OFFSET: usize = Self::BK_OFFSET + <Replicated<BK> as Serializable>::Size::USIZE;
+    const BTT_END: usize =
+        (Self::TV_OFFSET + <Replicated<TV> as Serializable>::Size::USIZE + 15) & (!15);
+
+    /// # Panics
+    /// If report length does not fit in `u16`.
+    pub fn encrypted_len(&self) -> u16 {
+        let len = EncryptedOprfReport::<BK, TV, TS, &[u8]>::SITE_DOMAIN_OFFSET
+            + self.site_domain.as_bytes().len();
+        len.try_into().unwrap()
+    }
+
+    /// # Errors
+    /// If there is a problem encrypting the report.
+    pub fn delimited_encrypt_to<R: CryptoRng + RngCore, B: BufMut>(
+        &self,
+        key_id: KeyIdentifier,
+        key_registry: &impl PublicKeyRegistry,
+        rng: &mut R,
+        out: &mut B,
+    ) -> Result<(), InvalidReportError> {
+        out.put_u16_le(self.encrypted_len());
+        self.encrypt_to(key_id, key_registry, rng, out)
+    }
+
+    /// # Errors
+    /// If there is a problem encrypting the report.
+    pub fn encrypt<R: CryptoRng + RngCore>(
+        &self,
+        key_id: KeyIdentifier,
+        key_registry: &impl PublicKeyRegistry,
+        rng: &mut R,
+    ) -> Result<Vec<u8>, InvalidReportError> {
+        let mut out = Vec::new();
```

Comment:
I missed it in the original, we could do `Vec::with_capacity(self.encrypted_len())` here to possibly avoid some reallocation overhead.